### PR TITLE
Move process image size to LoadInfo header

### PR DESCRIPTION
**Pull request by yuriks**
This removes a lot of hapzard offsetting from the process loader. Also clean up various part of the process loader.

### CODE REVIEW

File: src/main/main.rs [hunk 1]
```diff
@@ -27,7 +29,7 @@ pub extern fn main() {
 
 
     let processes = unsafe {
-        process::process::load_processes(&_sapps)
+        process::process::load_processes(&_sapps, &_eapps)
```
Comment (alevy) [inline src/main/main.rs]: Ah, this is actually incorrect (although there is no reasonable way to have known this from the code). &_eapps is only meaningful if the user-level processes are compiled into the kernel (which is true on some platforms but not generally, and definitely not on the Firestorm). Otherwise, processes are programmed onto flash out of band (on the Firestorm, via `tools/program/firestorm.py`) and the `.apps.*` sections have size 0 and `_eapps == _sapps`.

In this case, we know we've reached the last app to load because the length field is 0 (in practice we do this by just padding the last app with `0x000000`).

**Comment by alevy** [comment]
I left a comment about this inline, but expounding here since it maybe warrents a higher level discussion.

`_eapps` is actually meaningless right now. On platforms, like the Firestorm, where apps are programmed onto flash out-of-band from compiling the kernel (e.g. after the kernel is already flashed to the device), the `.apps.*` sections are empty and `_eapps == sapps`.

Instead, we rely on the last process in flash being padded with a `0x000000` word, thus resulting in a `length` zero, marking the end of the apps. On the Firestorm, `tools/program/firestorm.py` adds this word to the end of the concatenated app binaries. On other platforms where we _do_ compile apps into the kernel, we add that word using a hack in the linker script and chip crate (see https://github.com/alevy/tock-hotel/blob/master/src/chips/hotel/linker.ld#L15 and https://github.com/alevy/tock-hotel/blob/master/src/chips/hotel/lib.rs#L58).

This is all quite messy. Perhaps a better solution higher-level would be to let the platform decide how to "find" apps, and have the `main`/`process` crate simply expose `Process#create` as a mechanism for tell the core kernel where each successive app is. I don't think this is currently reasonable because the dependency is from `main` & `process` to `platform`, but I'm working on flipping that in this branch: https://github.com/alevy/tock/tree/build/dependencies (soon to be a pull request). Once we have that, I think we can rethink process loading in a more principled way.

In summary... for now, unfortunately I think you may only be able to do the simple cosmetic fix of moving the length word into the first field of `LoadInfo`.

**Comment by yuriks** [comment]
Ah, that explains why the padding 0 wasn't present in all platforms in the link scripts and why it seemed to be empty on some of the map files. I assumed that it was just sloppiness (as in, it just happens that the next word is usually a 0 by chance and it works out) but it's a fundamentally different approach. I should've asked you about it yesterday but it was late and I didn't want to lose momentum.

I'll adjust the code to actually watch for the sentinel for the time being.

**Comment by yuriks** [comment]
Updated. How does this look?

File: src/process/process.rs [hunk 2]
```diff
@@ -97,19 +98,30 @@ pub struct Process<'a> {
...
-        } else {
-            *op = None;
+    let mut addr = start_addr;
+    let mut process_iter = PROCS.iter_mut();
+
+    loop {
+        // The first member of the LoadInfo header contains the total size of each process image. A
+        // sentinel value of 0 (invalid because it's smaller than the header itself) is used to
+        // mark the end of the list of processes.
+        let total_size = *(addr as *const usize);
```
Comment (alevy) [inline src/process/process.rs]: Could actually already cast it to a `LoadInfo` here:

```rust
let load_info = &*(addr as *const LoadInfo);
if load_info.total_size == 0 {
    break;
}
```

Not a big difference for now, I think it might make this code slightly less magical. I suspect it'll become more useful to do this when this gets refactored, but just a weak suggestion for now.

File: src/process/process.rs [hunk 3]
```diff
@@ -97,19 +98,30 @@ pub struct Process<'a> {
...
+        if total_size == 0 {
+            break;
         }
+
+        let process = process_iter.next().expect("Exceeded maximum NUM_PROCS.");
```
Comment (alevy) [inline src/process/process.rs]: Hrm... My first intuition was that we shouldn't panic if there are more processes in flash than can fit in memory (just don't load them all), but for now (while we only load processes on boot anyway), it might better to have an early signal of failure then oops I loaded three apps and the third isn't working for some reason.

File: src/process/process.rs (hunk 2 above)
Comment (yuriks) [inline src/process/process.rs]: I avoided doing this because, technically, you're not supposed to access any of the other fields in the struct since they might be invalid memory. Practically speaking it doesn't make a difference but it feels like not using the struct is less "undefined behavior"-ish. (Avoiding this is why I changed it to use the pointer ranges in the previous version in the first place.)

File: src/process/process.rs [hunk 4]
```diff
@@ -351,35 +361,35 @@ struct LoadResult {
     app_mem_start: *const u8
 }
 
-unsafe fn load(mut start_addr: *const u8, mem_base: *mut u8) -> LoadResult {
+unsafe fn load(start_addr: *const u8, mem_base: *mut u8) -> LoadResult {
```
Comment (alevy) [inline src/process/process.rs]: Yeah this version is much clearer :+1:

File: src/process/process.rs [hunk 5]
```diff
@@ -63,8 +63,9 @@ pub struct Callback {
     pub pc: usize
 }
 
-#[repr(C,packed)]
```
Comment (alevy) [inline src/process/process.rs]: Why not `packed`? Since they are all `usize`s, it doesn't matter in practice, but aren't there no guarantees of the spacing between them otherwise? My concern is that we're generating the binary on a totally different platform than where we're reading it, so we have to be extra careful we know exactly how it gets laid out by the compiler.

File: src/process/process.rs (hunk 3 above)
Comment (yuriks) [inline src/process/process.rs]: My rationale for panicing is, apart from totally surprising, it could also be a security problem if one of those processes does some important task, fails to load and you don't notice it. Maybe have something like a priority list or a flag for essential processes which the system should panic if they couldn't be loaded. (And if you allow dynamic loading of processes there's a lot of latitude here for user-space based policy, so it might not be something to do in the kernel.)

**Comment by alevy** [comment]
Looks generally good. Will test it

File: src/process/process.rs (hunk 5 above)
Comment (yuriks) [inline src/process/process.rs]: The (Rustnomicon)[https://doc.rust-lang.org/nomicon/other-reprs.html] heavily discourages you from using `packed`. Basically, there's no guarantee that the platform will actually be able to work with a packed struct if it doesn't support unaligned accesses, and `repr(C)` already gives you enough guarantees that no extraneous padding will be introduced. I think if we wanted to be extra safe we could switch to using `u32` here instead, but as is we don't even make sure to deal with endianess when writing the struct so I think it's a moot point. Does this sound reasonable?

Comment (alevy) [inline src/process/process.rs]: Hrm... yes reasonable

Comment (ppannuto) [inline src/process/process.rs]: I'm not sure what the rust equivalent would be, but one habit from C that helps verify no funny padding business is going on:

    _Static_assert(sizeof(struct type_t) == 32)

File: src/process/process.rs (hunk 3 above)
Comment (alevy) [inline src/process/process.rs]: Yeah, since the specific choice of which processes to load and where to put them ought to be deffered to the platform anyway, I think you're right and this is the right choice for now

File: src/process/process.rs (hunk 5 above)
Comment (alevy) [inline src/process/process.rs]: yeah you can do that as a no-op with `transmute`. See the new impl for `static_init!`

**Comment by alevy** [comment]
Tested. It works

File: src/process/process.rs (hunk 5 above)
Comment (phil-levis) [inline src/process/process.rs]: Given that Tock is explicitly designed for Cortex Ms, I think it's OK to not worry about code being able to run on other architectures (except maybe Curie).
